filters/openpolicyagent/openpolicyagent.go

```diff
@@ -0,0 +1,318 @@
+}
+
+func WithPolicyType(policyType envoy.PolicyType) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.policyType = policyType
+		return nil
+	}
+}
+
+func WithConfigTemplate(configTemplate []byte) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.configTemplate = configTemplate
+		return nil
+	}
+}
+
+func WithConfigTemplateFile(configTemplateFile string) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		var err error
+		cfg.configTemplate, err = os.ReadFile(configTemplateFile)
+		return err
+	}
+}
+
+func WithEnvoyContextExtensions(envoyContextExtensions map[string]string) func(*OpenPolicyAgentInstanceConfig) error {
+	return func(cfg *OpenPolicyAgentInstanceConfig) error {
+		cfg.envoyContextExtensions = envoyContextExtensions
+		return nil
+	}
+}
+
+func (cfg *OpenPolicyAgentInstanceConfig) GetPolicyType() envoy.PolicyType {
+	return cfg.policyType
+}
+
+func (cfg *OpenPolicyAgentInstanceConfig) GetEnvoyContextExtensions() map[string]string {
+	return cfg.envoyContextExtensions
+}
+
+func NewOpenPolicyAgentConfig(opts ...func(*OpenPolicyAgentInstanceConfig) error) (*OpenPolicyAgentInstanceConfig, error) {
+	cfg := OpenPolicyAgentInstanceConfig{}
+
+	cfg.policyType = envoy.IngressPolicyType
+	if val, found := os.LookupEnv("OPA_ENVOY_POLICY_TYPE"); found {
+		cfg.policyType = envoy.PolicyType(val)
+	}
+
+	for _, opt := range opts {
+		if err := opt(&cfg); err != nil {
+			return nil, err
+		}
+	}
+
+	err := cfg.policyType.IsValid()
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.configTemplate == nil {
+		cfg.configTemplate, err = os.ReadFile(configTemplate())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &cfg, nil
+}
+
+func configTemplate() string {
+	var configTemplate string
+	var found bool
+	if configTemplate, found = os.LookupEnv("OPA_CONFIG_TEMPLATE_FILE"); !found {
+		configTemplate = "opaconfig.yaml"
+	}
+
+	return configTemplate
+}
+
+func (factory *OpenPolicyAgentFactory) BundleStorage() storage.Store {
+	if factory.bundleStorage == nil {
+		return inmem.New()
+	} else {
+		return factory.bundleStorage
+	}
+}
+
+func (factory *OpenPolicyAgentFactory) NewOpenPolicyAgentEnvoyInstance(bundleName string, config OpenPolicyAgentInstanceConfig, filterName string) (*OpenPolicyAgentInstance, error) {
+	runtime.RegisterPlugin(envoy.PluginName, envoy.Factory{})
+
+	configBytes, err := interpolateConfigTemplate(config.configTemplate, bundleName)
+	if err != nil {
+		return nil, err
+	}
+
+	engine, err := New(factory.BundleStorage(), configBytes, config, filterName, bundleName)
+
+	if err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+	if err = engine.Start(ctx); err != nil {
+		return nil, err
+	}
+
+	if err := engine.waitPluginsReady(
+		100*time.Millisecond,
+		time.Second*time.Duration(30)); err != nil {
```

Comment:
this reads a bit wild. Can you unfold the func please and maybe reuse `err`?
so 
```
err = engine.waitPluginsReady(100*time.Millisecond, time.Second*time.Duration(30))
if err != nil {...}
```